enhancements/proposed/20200501-json-output/start-json-output.md

```diff
@@ -0,0 +1,177 @@
+
+### Stdout
+As mentioned above, the requirements for stdout are:
+1. Steps are numbered, and the total number of steps is known
+1. Steps have a helpful name
+
+First, we'll need some way of distinguishing between logs.
+Each log has three components:
+1. The message itself 
+1. The emoji associated with the log (StyleEnum)
+1. Whether the log is of type Log, Warning, or Error
+
+```go
+type Log struct {
+	LogType // Will be either Log, Warning or Error
+	style   StyleEnum
+	message string
+}
+```
+
+
+We also need to know the total number of steps before minikube starts.
+For that to be possible, we need to know all of the logs we will print before starting.
+
+I propose creating a registry of logs, which is prefilled with all necessary logs before we start minikube.
+The registry would look like this:
+
+```go
+// Registry holds all user-facing logs
+type Registry struct {
+	// maps the name of the log to a Log type
+	Logs  map[string]Log
+	Index int
+}
+```
+
+at the beginning of `minikube start`, we would initialize a `Registry` type in the `out` package, which would exist as a global variable.
+All logs would be added to the registry at this time via `Register`:
+
+```go
+// Register registers a log
+func Register(name string, style StyleEnum, message string, logType LogType) {
+	registry.Logs[name] = log{
+		style:   style,
+		message: message,
+		LogType: logType,
+	}
+}
+```
+
+and before starting minikube we would run `initializeRegistry()`, which will register all logs we expect to output:
+
+```go
+func initializeRegistry()
+    Register("select_driver", out.Sparkle, `Using the {{.driver}} driver based on existing profile`, Log)
+    // Add all other logs here as well
+
+```
+
+This log would be printed at the correct time later in the code by calling:
+
+```go
+func Print(name string, a ...V)
+```
+
+which would find the correct log in the registry and apply the template to it.
+
+In the code itself, this means that this line, which currently looks like this:
+
+```go
+out.T(out.Sparkle, `Using the {{.driver}} driver based on existing profile`, out.V{"driver": ds.String()})
+```
+
+would now be:
+
+```go
+out.T("select_driver", out.V{"driver": ds.String()})
+```
+
+We can use the `Index` field in `Registry` to track which number log we are currently at. 
+Since logs have been pre-registered, we know what the total number of expected logs is.
+
+Similarly to stderr, if the JSON flag is specified, we will print the JSON encoding of the `Log` struct to stdout instead of the expected log in `out.T`, `out.Warning` and `out.Err`.
+
+
+#### Testing Plan
+Both unit tests and integration tests will be required to test this feature.
+
+Unit tests will cover:
+1. That the JSON output is correct and parsable
+
+Integration tests will cover:
+1. That all output to stdout is in JSON, ensuring that all user-facing logs have been registered
+   
+
```

Comment:
Can you share how an example `minikube start` run would be rendered to stdout/stderr? It would be helpful to understand if the new outputs match what users are looking for.